server/src/main/java/org/elasticsearch/index/mapper/KeyedJsonAtomicFieldData.java

```diff
@@ -0,0 +1,230 @@
+import org.elasticsearch.index.fielddata.FieldData;
+import org.elasticsearch.index.fielddata.ScriptDocValues;
+import org.elasticsearch.index.fielddata.SortedBinaryDocValues;
+import org.elasticsearch.index.fielddata.plain.AbstractAtomicOrdinalsFieldData;
+
+import java.io.IOException;
+import java.util.Collection;
+
+/**
+ * The atomic field data implementation for {@link JsonFieldMapper.KeyedJsonFieldType}.
+ *
+ * This class wraps the field data that is built directly on the keyed JSON field,
+ * and filters out values whose prefix doesn't match the requested key.
+ */
+public class KeyedJsonAtomicFieldData implements AtomicOrdinalsFieldData {
+
+    private final String key;
+    private final AtomicOrdinalsFieldData delegate;
+
+    KeyedJsonAtomicFieldData(String key,
+                             AtomicOrdinalsFieldData delegate) {
+        this.key = key;
+        this.delegate = delegate;
+    }
+
+    @Override
+    public long ramBytesUsed() {
+        return delegate.ramBytesUsed();
+    }
+
+    @Override
+    public Collection<Accountable> getChildResources() {
+        return delegate.getChildResources();
+    }
+
+    @Override
+    public SortedSetDocValues getOrdinalsValues() {
+        SortedSetDocValues values = delegate.getOrdinalsValues();
+        return KeyedJsonDocValues.create(key, values);
+    }
+
+    @Override
+    public void close() {
+        delegate.close();
+    }
+
+    @Override
+    public ScriptDocValues<?> getScriptValues() {
+        return AbstractAtomicOrdinalsFieldData.DEFAULT_SCRIPT_FUNCTION
+            .apply(getOrdinalsValues());
+    }
+
+    @Override
+    public SortedBinaryDocValues getBytesValues() {
+        return FieldData.toString(getOrdinalsValues());
+    }
+
+    static class KeyedJsonDocValues extends AbstractSortedSetDocValues {
+
+        private final BytesRef key;
+        private final SortedSetDocValues delegate;
+
+        /**
+         * The first and last ordinals whose term has 'key' as a prefix.
+         */
+        private final long minOrd;
+        private final long maxOrd;
+
+        public static KeyedJsonDocValues create(String key, SortedSetDocValues delegate) {
+            BytesRef keyBytes = new BytesRef(key);
+
+            long minOrd, maxOrd;
+            try {
+                minOrd = findMinOrd(keyBytes, delegate);
+                maxOrd = minOrd != -1 ? findMaxOrd(keyBytes, delegate) : -1;
+            } catch (IOException e) {
+                throw new RuntimeException(e);
+            }
+
+            return new KeyedJsonDocValues(keyBytes, delegate, minOrd, maxOrd);
+        }
+
+        private KeyedJsonDocValues(BytesRef key,
+                                   SortedSetDocValues delegate,
+                                   long minOrd,
+                                   long maxOrd) {
+            this.key = key;
+            this.delegate = delegate;
+            this.minOrd = minOrd;
+            this.maxOrd = maxOrd;
```

Comment:
One small optimization that could save some memory in the `terms` aggregation is to remap the documents ordinals to `[0, (maxOrd-minOrd)]` during the collection. With global ordinals the `terms` aggregation allocates one big array based on the size reported by `getValueCount` so if you have a lot inner fields with lots of different values we'll allocate much more than what we really need. The remapping should happen in `lookupOrd` (to retrieve the original ordinal) and `nextOrd` (to remap based on minOrd), `getValueCount` can return `maxOrd-minOrd`.

Reply:
+1 We might need to keep it for a follow-up however because I think it's going to make IndexOrdinalsFieldData#getOrdinalMap a bit hard to implement given that OrdinalMap is not designed for being extended.